File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int maxLIS = 1, n = nums.size();

        vector<int> memo(n, -1);
        
        for (int i = n - 1; i >= 0; --i) {
            int lis = 1;
            for (int j = i + 1; j < n; ++j)
                if (nums[j] > nums[i])
                    lis = max(lis, memo[j] + 1);
            memo[i] = lis;
            maxLIS = max(lis, maxLIS);
        }
        return maxLIS;
    }
};

File: Data Structures & Algorithms/house-robber/submission-4.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n, nn = 0, 0
        for i in range(len(nums) - 1, -1, -1):
            c = max(nums[i] + nn, n)
            nn = n
            n = c
        return c
